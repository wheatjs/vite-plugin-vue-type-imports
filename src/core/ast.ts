import fs from 'fs'
import type {
  CallExpression,
  ExportNamedDeclaration,
  ImportDeclaration,
  Node,
  Program,
  StringLiteral,
  TSEnumDeclaration,
  TSExpressionWithTypeArguments,
  TSInterfaceDeclaration,
  TSTypeAliasDeclaration,
  TSTypeLiteral,
  TSTypeParameterInstantiation,
  TSTypeReference,
  TSUnionType,
} from '@babel/types'
import type {
  MaybeAliases,
  Replacement,
  StringMap,
} from './utils'
import {
  getAst,
  groupImports,
  insertString,
  intersect,
  resolveModulePath,
} from './utils'

const DEFINE_PROPS = 'defineProps'
const DEFINE_EMITS = 'defineEmits'
const WITH_DEFAULTS = 'withDefaults'
const TS_TYPES_KEYS = ['TSTypeAliasDeclaration', 'TSInterfaceDeclaration', 'TSEnumDeclaration']

const isDefineProps = (node: Node): node is CallExpression => isCallOf(node, DEFINE_PROPS)
const isDefineEmits = (node: Node): node is CallExpression => isCallOf(node, DEFINE_EMITS)
const isWithDefaults = (node: Node): node is CallExpression => isCallOf(node, WITH_DEFAULTS)

export interface IImport {
  start: number
  end: number
  local: string
  imported: string
  path: string
}

export interface InterfaceMetaData {
  extendInterfaceName?: string
  interfaceBodyStart?: number
  isProperty?: boolean
}

export type MaybeNumber = number | null | undefined

export type MaybeNode = Node | null | undefined

export type ExportNamedFromDeclaration = ExportNamedDeclaration & { source: StringLiteral }

export type TypeInfo = Partial<Record<'type' | 'name', string>>

export type GetTypesResult = (string | TypeInfo)[]

export interface GetImportsResult {
  imports: IImport[]
  importNodes: ImportDeclaration[]
}

export type TSTypes = TSTypeAliasDeclaration | TSInterfaceDeclaration | TSEnumDeclaration

export type NodeMap = Map<string, TSTypes>

export function getAvailableImportsFromAst(ast: Program): GetImportsResult {
  const imports: IImport[] = []
  const importNodes: ImportDeclaration[] = []

  const addImport = (node: ImportDeclaration) => {
    for (const specifier of node.specifiers) {
      if (specifier.type === 'ImportSpecifier' && specifier.imported.type === 'Identifier') {
        imports.push({
          start: specifier.imported.start!,
          end: specifier.local.end!,
          imported: specifier.imported.name,
          local: specifier.local.name,
          path: node.source.value,
        })
      }
    }

    importNodes.push(node)
  }

  for (const node of ast.body) {
    if (node.type === 'ImportDeclaration' && node.specifiers.length)
      addImport(node)
  }

  return { imports, importNodes }
}

/**
 * get reExported fields
 *
 * e.g. export { x } from './xxx'
 */
export function getAvailableExportsFromAst(ast: Program) {
  const exports: IImport[] = []

  // TODO: Support 'import { a as b }' syntax
  const addExport = (node: ExportNamedFromDeclaration) => {
    for (const specifier of node.specifiers) {
      if (specifier.type === 'ExportSpecifier' && specifier.exported.type === 'Identifier') {
        exports.push({
          start: specifier.exported.start!,
          end: specifier.local.end!,
          imported: specifier.exported.name,
          local: specifier.local.name,
          path: node.source.value,
        })
      }
    }
  }

  for (const node of ast.body) {
    // TODO: support export * from
    if (isExportNamedFromDeclaration(node))
      addExport(node)
  }

  return exports
}

export function getUsedInterfacesFromAst(ast: Program) {
  const interfaces: string[] = []

  const addInterface = (node: Node) => {
    if (node.type === 'CallExpression' && node.typeParameters?.type === 'TSTypeParameterInstantiation') {
      const propsTypeDefinition = node.typeParameters.params[0]

      if (propsTypeDefinition.type === 'TSTypeReference' && propsTypeDefinition.typeName.type === 'Identifier')
        interfaces.push(propsTypeDefinition.typeName.name)

      // TODO: Support nested type params
      // if (propsTypeDefinition.typeParameters)
      //     interfaces.push(...getTypesFromTypeParameters(propsTypeDefinition.typeParameters));
    }
  }

  for (const node of ast.body) {
    if (node.type === 'ExpressionStatement') {
      if (isWithDefaults(node.expression))
        addInterface(node.expression.arguments[0])
      else if (isDefineProps(node.expression) || isDefineEmits(node.expression))
        addInterface(node.expression)
    }

    if (node.type === 'VariableDeclaration' && !node.declare) {
      for (const decl of node.declarations) {
        if (decl.init) {
          if (isWithDefaults(decl.init))
            addInterface(decl.init.arguments[0])
          else if (isDefineProps(decl.init) || isDefineEmits(decl.init))
            addInterface(decl.init)
        }
      }
    }
  }

  return interfaces
}

function getTypesFromTypeParameters(x: TSTypeParameterInstantiation) {
  const types: GetTypesResult = []

  for (const p of x.params) {
    if (p.type === 'TSTypeLiteral') { types.push(...getTSTypeLiteralTypes(p)) }
    else if (p.type === 'TSTypeReference') {
      if (p.typeName.type === 'Identifier')
        types.push(p.typeName.name)
    }
  }

  return types
}

function getTSTypeLiteralTypes(x: TSTypeLiteral) {
  const types: GetTypesResult = []

  for (const m of x.members) {
    if (m.type === 'TSPropertySignature') {
      if (m.typeAnnotation?.typeAnnotation.type === 'TSTypeLiteral') {
        types.push(...getTSTypeLiteralTypes(m.typeAnnotation.typeAnnotation))
      }
      else if (m.typeAnnotation?.typeAnnotation.type === 'TSTypeReference') {
        if (m.typeAnnotation.typeAnnotation.typeName.type === 'Identifier') {
          // TODO: understand why we push a object
          types.push({
            type: m.typeAnnotation.typeAnnotation.type,
            name: m.typeAnnotation.typeAnnotation.typeName.name,
          })
        }

        if (m.typeAnnotation.typeAnnotation.typeParameters)
          types.push(...getTypesFromTypeParameters(m.typeAnnotation.typeAnnotation.typeParameters))
      }
      else {
        types.push({ type: m.typeAnnotation?.typeAnnotation.type })
      }
    }
  }

  return types
}

function extractAllTypescriptTypesFromAST(ast: Program, isInternal: boolean) {
  return ast.body
    .map((node) => {
      // e.g. 'export interface | type | enum'
      if (node.type === 'ExportNamedDeclaration' && node.declaration && isTSTypes(node.declaration, isInternal))
        return node.declaration

      // e.g. 'interface | type | enum'
      if (isTSTypes(node, isInternal))
        return node

      return null
    })
    .filter((x): x is TSTypes => x !== null)
}

type ExtractedTypes = StringMap
type MetaDataMap = Map<string, InterfaceMetaData>

interface ExtractTypesFromSourceOptions {
  relativePath: string
  aliases: MaybeAliases
  extractedTypes?: ExtractedTypes
  metaDataMap?: MetaDataMap
  // For internal interfaces
  ast?: Program
  isInternal?: boolean
  cleanInterface?: boolean
}

interface ExtractResult {
  result: StringMap
  importNodes: ImportDeclaration[]
  extraSpecifiers: string[]
  extraReplacements: Replacement[]
}

/**
 * Given a specific source file, extract the specified types.
 */
export async function extractTypesFromSource(
  source: string,
  types: string[],
  options: ExtractTypesFromSourceOptions,
): Promise<ExtractResult> {
  const {
    relativePath,
    aliases,
    ast = getAst(source),
    isInternal = false,
    cleanInterface = false,
    extractedTypes = new Map<string, string>(),
    metaDataMap = new Map<string, InterfaceMetaData>(),
  } = options

  // console.log(relativePath);

  const missingTypes: string[] = []

  // Get external types
  const { imports, importNodes } = getAvailableImportsFromAst(ast)

  if (!isInternal)
    imports.push(...getAvailableExportsFromAst(ast))

  // TODO: Fix duplicate key name
  const nodeMap = getTSNodeMap(extractAllTypescriptTypesFromAST(ast, isInternal))

  const extraSpecifiers: string[] = []
  const extraReplacements: Replacement[] = []

  const extractFromPosition = (start: MaybeNumber, end: MaybeNumber) =>
    isNumber(start) && isNumber(end) ? source.slice(start, end) : ''

  function removeInterface(node: TSInterfaceDeclaration) {
    extraReplacements.push({
      start: node.start!,
      end: node.end!,
      replacement: '',
    })
  }

  function getTSNodeMap(nodes: TSTypes[]): NodeMap {
    const nodeMap = new Map<string, TSTypes>()

    for (const node of nodes) {
      if ('name' in node.id)
        nodeMap.set(node.id.name, node)
    }

    return nodeMap
  }

  function ExtractTypeByNode(node: TSTypes) {
    switch (node.type) {
      // Types e.g. export Type Color = 'red' | 'blue'
      case 'TSTypeAliasDeclaration': {
        extractTypesFromTypeAlias(node)
        break
      }
      // Interfaces e.g. export interface MyInterface {}
      case 'TSInterfaceDeclaration': {
        extractTypesFromInterface(node)
        break
      }
      // Enums e.g. export enum UserType {}
      case 'TSEnumDeclaration': {
        extractTypesFromEnum(node)
        break
      }
    }
  }

  /**
   * Extract ts types by name.
   */
  function extractTypeByName(name: string) {
    // Skip already extracted types
    if (extractedTypes.get(name))
      return

    const node = nodeMap.get(name)

    if (node)
      ExtractTypeByNode(node)
    else
      missingTypes.push(name)
  }

  // Recursively calls this function to find types from other modules.
  const extractTypesFromModule = async (modulePath: string, types: string[]) => {
    const path = await resolveModulePath(modulePath, relativePath, aliases)

    if (!path)
      return

    // NOTE: Slow when use fsPromises.readFile(), tested on Arch Linux x64 (Kernel 5.16.11)
    // Wondering what make it slow. Temporarily, use fs.readFileSync() instead.
    const contents = fs.readFileSync(path, 'utf-8')

    await extractTypesFromSource(contents, types, {
      relativePath: path,
      aliases,
      extractedTypes,
      metaDataMap,
    })
  }

  const extractTypesFromTSUnionType = (union: TSUnionType) => {
    union.types
      .filter((n): n is TSTypeReference => n.type === 'TSTypeReference')
      .forEach((typeReference) => {
        if (typeReference.typeName.type === 'Identifier')
          extractTypeByName(typeReference.typeName.name)
      })
  }

  function extractExtendInterfaces(interfaces: TSExpressionWithTypeArguments[], interfaceMetaData: InterfaceMetaData) {
    for (const extend of interfaces) {
      if (extend.expression.type === 'Identifier') {
        const name = extend.expression.name
        metaDataMap.set(name, interfaceMetaData)

        if (isInternal)
          extraSpecifiers.push(name)

        extractTypeByName(name)
      }
    }
  }

  /**
   * Extract ts type interfaces. Should also check top-level properties
   * in the interface to look for types to extract
   */
  const extractTypesFromInterface = (node: TSInterfaceDeclaration) => {
    const { extendInterfaceName, interfaceBodyStart, isProperty } = metaDataMap.get(node.id.name) ?? {}

    const interfaceName = node.id.name
    const extendsInterfaces = node.extends

    // Skip all process, since Vue only transform the type of nested objects to 'Object'
    if (isProperty) {
      extractedTypes.set(interfaceName, `interface ${interfaceName} {}`)
      return
    }

    const bodyStart = node.body.start!
    const bodyEnd = node.body.end!

    if (extendInterfaceName) {
      extractedTypes.set(
        extendInterfaceName,
        insertString(
          extractedTypes.get(extendInterfaceName)!,
          interfaceBodyStart! + 1,
          extractFromPosition(bodyStart + 1, bodyEnd - 1),
        ),
      )

      if (isInternal && cleanInterface)
        removeInterface(node)

      if (extendsInterfaces) {
        extractExtendInterfaces(extendsInterfaces, {
          extendInterfaceName,
          interfaceBodyStart: interfaceBodyStart!,
        })
      }
    }
    else {
      extractedTypes.set(interfaceName, `interface ${interfaceName} ${extractFromPosition(bodyStart, bodyEnd)}`)

      if (extendsInterfaces) {
        if (isInternal)
          removeInterface(node)

        extractExtendInterfaces(extendsInterfaces, {
          extendInterfaceName: interfaceName,
          // 'interface A '.length -> 12
          interfaceBodyStart: interfaceName.length + 11,
        })
      }
      // No need to extract an individual interface
      else if (isInternal) {
        extractedTypes.delete(interfaceName)
      }
    }

    for (const prop of node.body.body) {
      if (prop.type === 'TSPropertySignature') {
        // TODO: Should this be filtered?
        if (prop.typeAnnotation?.typeAnnotation.type === 'TSUnionType') {
          extractTypesFromTSUnionType(prop.typeAnnotation.typeAnnotation)
        }
        else if (
          prop.typeAnnotation?.typeAnnotation.type === 'TSTypeReference'
          && prop.typeAnnotation.typeAnnotation.typeName.type === 'Identifier'
        ) {
          metaDataMap.set(prop.typeAnnotation.typeAnnotation.typeName.name, { isProperty: true })
          extractTypeByName(prop.typeAnnotation.typeAnnotation.typeName.name)
        }
      }
    }
  }

  /**
   * Extract types from TSTypeAlias
   */
  const extractTypesFromTypeAlias = (node: TSTypeAliasDeclaration) => {
    extractedTypes.set(node.id.name, extractFromPosition(node.start, node.end))

    if (node.typeAnnotation.type === 'TSUnionType')
      extractTypesFromTSUnionType(node.typeAnnotation)

    // TODO: Support TSLiteral, IntersectionType
    if (node.typeAnnotation.type === 'TSTypeReference' && node.typeAnnotation.typeName.type === 'Identifier')
      extractTypeByName(node.typeAnnotation.typeName.name)
  }

  /**
   * Extract enum types. Since I don't believe these can depend on any other
   * types we just want to extract the string itself.
   *
   * Zorin: Since Vue can't handle Enum types right now, would it be better to convert it to 'type [name] = number | string;'?
   */
  const extractTypesFromEnum = (node: TSEnumDeclaration) => {
    const enumName = node.id.name
    const enumTypes: Set<string> = new Set()

    // (semi-stable) Determine the type of enum, may not be able to process the use of complex scenes
    for (const member of node.members) {
      if (member.initializer) {
        if (member.initializer.type === 'NumericLiteral')
          enumTypes.add('number')
        else if (member.initializer.type === 'StringLiteral')
          enumTypes.add('string')
      }
      else {
        enumTypes.add('number')
      }
    }

    extractedTypes.set(enumName, `type ${enumName} = ${[...enumTypes].join(' | ') || 'number | string'};`)
  }

  for (const typeName of types)
    extractTypeByName(typeName)

  if (missingTypes.length) {
    await Promise.all(
      Object.entries(groupImports(imports)).map(async ([modulePath, importedFields]) => {
        const intersection = intersect(importedFields, missingTypes)

        if (intersection.length)
          await extractTypesFromModule(modulePath, intersection)
      }),
    )
  }

  return {
    result: extractedTypes,
    importNodes,
    extraSpecifiers,
    extraReplacements,
  }
}

export function isNumber(n: MaybeNumber): n is number {
  return typeof n === 'number'
}

export function isCallOf(node: MaybeNode, test: string | ((id: string) => boolean)): node is CallExpression {
  return !!(
    node
    && node.type === 'CallExpression'
    && node.callee.type === 'Identifier'
    && (typeof test === 'string' ? node.callee.name === test : test(node.callee.name))
  )
}

export function isTSTypes(node: MaybeNode, isInternal: boolean): node is TSTypes {
  if (isInternal)
    return !!(node && node.type === 'TSInterfaceDeclaration')

  return !!(node && TS_TYPES_KEYS.includes(node.type))
}

export function isExportNamedFromDeclaration(node: MaybeNode): node is ExportNamedFromDeclaration {
  return !!(node && node.type === 'ExportNamedDeclaration' && node.source)
}
