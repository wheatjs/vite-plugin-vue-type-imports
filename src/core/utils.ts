import { existsSync, readFileSync } from 'fs'
import { dirname, extname, join } from 'path'
import fg from 'fast-glob'
import { resolveModule } from 'local-pkg'
import type { Alias, AliasOptions } from 'vite'
import { babelParse } from '@vue/compiler-sfc'
import type { Program } from '@babel/types'
import type { IImport } from './ast'

type Pkg = Partial<Record<'types' | 'typings', string>>

export type StringMap = Map<string, string>

export type MaybeAliases = ((AliasOptions | undefined) & Alias[]) | undefined

export function getAst(content: string): Program {
  return babelParse(content, {
    sourceType: 'module',
    plugins: ['typescript', 'topLevelAwait'],
  }).program
}

/**
 * Source: https://github.com/rollup/plugins/blob/master/packages/alias/src/index.ts
 */
export function matches(pattern: string | RegExp, importee: string) {
  if (pattern instanceof RegExp)
    return pattern.test(importee)

  if (importee.length < pattern.length)
    return false

  if (importee === pattern)
    return true

  const importeeStartsWithKey = importee.indexOf(pattern) === 0
  const importeeHasSlashAfterKey = importee.slice(pattern.length)[0] === '/'
  return importeeStartsWithKey && importeeHasSlashAfterKey
}

// https://github.com/antfu/local-pkg/blob/main/index.mjs
export function searchPackageJSON(dir: string): string | undefined {
  let packageJsonPath
  while (true) {
    if (!dir)
      return
    const newDir = dirname(dir)
    if (newDir === dir)
      return

    dir = newDir
    packageJsonPath = join(dir, 'package.json')
    if (existsSync(packageJsonPath))
      break
  }

  return packageJsonPath
}

export function resolvePath(path: string, from: string, aliases: MaybeAliases) {
  const matchedEntry = aliases?.find(entry => matches(entry.find, path))

  // Path which is using aliases. e.g. '~/types'
  if (matchedEntry)
    return path.replace(matchedEntry.find, matchedEntry.replacement)

  // External package
  const resolved_path = resolveModule(path)

  // Not a package. e.g. '../types'
  if (!resolved_path)
    return join(dirname(from), path)

  // Result is a typescript file. e.g. 'vue/macros-global.d.ts'
  if (extname(resolved_path) === '.ts') {
    return resolved_path
  }
  // Not a typescript file, find declaration file
  // The only situation is that the types are imported from the main entry. e.g. 'vue' -> 'vue/dist/vue.d.ts'
  else {
    const packageJsonPath = searchPackageJSON(resolved_path)

    if (!packageJsonPath)
      return

    const { types, typings } = JSON.parse(readFileSync(packageJsonPath, 'utf-8')) as Pkg

    let result: string | undefined

    try {
      // @ts-expect-error allow result to be undefined
      result = join(dirname(packageJsonPath), types || typings)
    }
    catch {}

    return result
  }
}

export async function resolveModulePath(path: string, from: string, aliases: MaybeAliases) {
  const maybePath = resolvePath(path, from, aliases)?.replace(/\\/g, '/')

  if (!maybePath)
    return null

  const files = await fg([`${maybePath}`, `${maybePath}*.+(ts|d.ts)`, `${maybePath}*/index.+(ts|d.ts)`], {
    onlyFiles: true,
  })

  if (files.length)
    return files[0]

  return null
}

/**
 * @returns Record<string, string[]> - key: the imported file, value: imported fields
 */
export function groupImports(imports: IImport[]) {
  return imports.reduce<Record<string, string[]>>((obj, importInfo) => {
    obj[importInfo.path] = obj[importInfo.path] || []
    obj[importInfo.path].push(importInfo.imported)

    return obj
  }, {})
}

export function intersect<A = any, B = any>(a: Array<A>, b: Array<B>): (A | B)[] {
  const setB = new Set(b)
  // @ts-expect-error unnecessary type checking (for now)
  return [...new Set(a)].filter(x => setB.has(x))
}

export interface Replacement {
  start: number
  end: number
  replacement: string
}

/**
 * Replace all items at specified indexes from the bottom up.
 */
export function replaceAtIndexes(source: string, replacements: Replacement[], clean = false): string {
  replacements.sort((a, b) => b.start - a.start)
  let result = source

  for (const node of replacements)
    result = result.slice(0, node.start) + node.replacement + result.slice(node.end)

  // remove empty newline -> ''
  if (clean) {
    result = result
      .split('\n')
      .filter(val => val)
      .join('\n')
  }

  return result
}

export function insertString(source: string, start: number, insertVal: string): string {
  return source.slice(0, start) + insertVal + source.slice(start)
}
