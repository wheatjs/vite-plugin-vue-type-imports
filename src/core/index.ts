import { parse } from '@vue/compiler-sfc';
import { Alias, AliasOptions } from 'vite';
import { extractTypesFromSource, getUsedInterfacesFromAst } from './ast';
import { getAst, MaybeAliases, replaceAtIndexes } from './utils';

export interface CleanOptions {
  newline?: boolean;
  interface?: boolean;
}

export interface TransformOptions {
  id: string;
  clean: CleanOptions;
  aliases?: MaybeAliases;
}

export async function transform(code: string, { id, aliases, clean }: TransformOptions) {
  const {
    descriptor: { scriptSetup },
  } = parse(code);

  if (scriptSetup?.lang !== 'ts' || !scriptSetup.content) return code;

  const program = getAst(scriptSetup.content);

  const interfaces = getUsedInterfacesFromAst(program);

  const { result, importNodes, extraSpecifiers, extraReplacements } = await extractTypesFromSource(
    scriptSetup.content,
    interfaces,
    {
      aliases,
      relativePath: id,
      ast: program,
      isInternal: true,
      cleanInterface: clean.interface,
    },
  );

  // Skip
  if (!result.size) {
    return code;
  }

  const resolvedTypes = [...result].reverse();
  const replacements = extraReplacements;

  // Clean up imports
  importNodes.forEach((i) => {
    const firstStart = i.specifiers[0].start!;
    const lastEnd = i.specifiers[i.specifiers.length - 1].end!;

    const savedSpecifiers = i.specifiers
      .map((specifier) => {
        if (specifier.type === 'ImportSpecifier' && specifier.imported.type === 'Identifier') {
          const name = specifier.imported.name;
          const shouldSave = !resolvedTypes.some((x) => x[0] === name);

          if (shouldSave && !extraSpecifiers.includes(name)) {
            return name;
          }

          return null;
        }

        return null;
      })
      .filter((s): s is string => s !== null);

    // Clean the whole import statement if no specifiers are saved.
    if (!savedSpecifiers.length) {
      replacements.push({
        start: i.start!,
        end: i.end!,
        replacement: '',
      });
    } else {
      replacements.push({
        start: firstStart,
        end: lastEnd,
        replacement: savedSpecifiers.join(', '),
      });
    }
  });

  const inlinedTypes = resolvedTypes.map((x) => x[1]).join('\n');

  const transformedCode = [
    // Tag head
    code.slice(0, scriptSetup.loc.start.offset),
    // Script setup content
    inlinedTypes,
    // Replace import statements
    replaceAtIndexes(scriptSetup.content, replacements, clean.newline),
    // Tag end
    code.slice(scriptSetup.loc.end.offset),
  ].join('\n');

  return transformedCode;
}
