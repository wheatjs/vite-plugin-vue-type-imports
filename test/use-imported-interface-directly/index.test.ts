import { transform, TransformOptions } from '~/core'

function generateCode(testPath: string) {
    return `
    <script lang="ts" setup>
    import { Props } from './${testPath}'
    
    defineProps<Props>()
    </script>
    `
}

const options: TransformOptions = { id: __filename, clean: {} }

test('Redeclaration of types #6', async () => {
    const result = await transform(generateCode('redeclaration-of-types'), options)

    expect(result).toMatchSnapshot()
});

test('Interface which has no references', async () => {
    const result = await transform(generateCode('no-reference-interface'), options)

    expect(result).toMatchSnapshot()
});

test('Interface extends interface (No references)', async () => {
    const result = await transform(generateCode('interface-extends-interface'), options)

    expect(result).toMatchSnapshot()
});
