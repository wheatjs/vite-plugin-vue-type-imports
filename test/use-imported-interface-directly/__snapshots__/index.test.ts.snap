// Vitest Snapshot v1

exports[`Interface extends interface (No references) 1`] = `
"
    <script lang=\\"ts\\" setup>
interface Props {
    baz: boolean;

    foo: string;
    bar: number;
}

    
    
    defineProps<Props>()
    
</script>
    "
`;

exports[`Interface which has no references 1`] = `
"
    <script lang=\\"ts\\" setup>
interface Props {
    foo: string;
    bar: number;
}

    
    
    defineProps<Props>()
    
</script>
    "
`;

exports[`Redeclaration of types #6 1`] = `
"
    <script lang=\\"ts\\" setup>
type Foo = [number, number]
interface Props {
    foo: Foo
    bar: Foo
}

    
    
    defineProps<Props>()
    
</script>
    "
`;
